File: packages/sdk/examples/exampleBrlOfframp.ts
import * as readline from "readline";
import { EPaymentMethod, EvmToken, FiatToken, Networks, RampDirection } from "../src/index";
import { VortexSdkConfig } from "../src/types";
import { VortexSdk } from "../src/VortexSdk";

function requireEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Missing required env var: ${name}. Copy .env.example to .env and set it.`);
  }
  return value;
}

async function runBrlOfframpExample() {
  const askQuestion = (query: string): Promise<string> => {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    return new Promise(resolve => {
      rl.question(query, (ans: string) => {
        rl.close();
        resolve(ans.trim());
      });
    });
  };

  try {
    console.log("Starting BRL Offramp Example...\n");

    console.log("📝 Step 1: Initializing VortexSdk...");
    const config: VortexSdkConfig = {
      apiBaseUrl: process.env.VORTEX_API_URL ?? "http://localhost:3000",
      autoReconnect: true,
      // Optional: provide custom WebSocket URLs
      moonbeamWsUrl: undefined,
      pendulumWsUrl: undefined, // 'wss://custom-moonbeam-rpc.com',
      publicKey: requireEnv("VORTEX_PUBLIC_KEY"), // 'wss://custom-pendulum-rpc.com',
      secretKey: requireEnv("VORTEX_SECRET_KEY"), // default is `true`
      // Optional: store ephemeral keys for debug
      storeEphemeralKeys: true // default is `true`
    };

    const sdk = new VortexSdk(config);
    console.log("⏳ Waiting for API initialization...");
    console.log("✅ VortexSdk initialized successfully\n");

    console.log("📝 Step 2: Creating quote for BRL offramp...");
    const quoteRequest = {
      from: Networks.Polygon,
      inputAmount: "100",
      inputCurrency: EvmToken.USDC,
      network: Networks.Polygon,
      outputCurrency: FiatToken.BRL,
      rampType: RampDirection.SELL,
      to: EPaymentMethod.PIX
    };

    const quote = await sdk.createQuote(quoteRequest);
    console.log("✅ Quote created successfully:");
    console.log(`   Quote ID: ${quote.id}`);
    console.log(`   Input: ${quote.inputAmount} ${quote.inputCurrency}`);
    console.log(`   Output: ${quote.outputAmount} ${quote.outputCurrency}`);
    console.log(`   Total Fee: ${quote.totalFeeFiat} ${quote.feeCurrency}`);
    console.log(`   Expires at: ${quote.expiresAt}\n`);

    const brlOfframpData = {
      pixDestination: requireEnv("BRL_PIX_DESTINATION"),
      receiverTaxId: requireEnv("BRL_RECEIVER_TAX_ID"),
      taxId: requireEnv("BRL_TAX_ID"),
      walletAddress: requireEnv("WALLET_ADDRESS")
    };

    const { rampProcess, unsignedTransactions } = await sdk.registerRamp(quote, brlOfframpData);

    console.log("✅ BRL Offramp registered successfully:");
    console.log(`   Ramp ID: ${rampProcess.id}`);

    // The unsignedTransactions object will always return the transactions the user must sign and broadcast, please check the docs for more information https://api-docs.vortexfinance.co/vortex-sdk-1289458m0.
    console.log("   Unsigned transactions:");
    unsignedTransactions.forEach(tx => {
      const { to, data, value } = sdk.getTransactionToBroadcast(tx);
      console.log(`     - ${tx.phase}: Send to ${to} data ${data} with value ${value}`);
    });
    console.log("");

    console.log(
      "\n🛑 Complete the token payment on-chain now. Execute the transactions shown above (squidRouterApprove and squidRouterSwap), and save the corresponding transaction hashes."
    );

    const squidRouterApproveHash = await askQuestion("➡️  Enter the Squid Router Approve Hash: ");
    const squidRouterSwapHash = await askQuestion("➡️  Enter the Squid Router Swap Hash: ");

    console.log("\n📝 Step 4: Updating BRL offramp...");
    const transactionHashes = {
      squidRouterApproveHash,
      squidRouterSwapHash
    };

    await sdk.updateRamp(quote, rampProcess.id, transactionHashes);
    console.log("✅ BRL Offramp updated successfully.");

    await sdk.startRamp(rampProcess.id);
    console.log("✅ Offramp started successfully.");
  } catch (error) {
    console.error("❌ Error in BRL Offramp Example:", error);
    if (error instanceof Error) {
      console.error("Error message:", error.message);
    }
    process.exit(1);
  }
}

if (require.main === module) {
  runBrlOfframpExample()
    .then(() => {
      console.log("\n✨ Example execution completed");
      process.exit(0);
    })
    .catch(error => {
      console.error("\n💥 Example execution failed:", error);
      process.exit(1);
    });
}

File: python-sdk/examples/example_brl_offramp.py
"""
Example: Brazilian Real (BRL) Offramp

This example demonstrates how to create a BRL offramp (crypto to PIX) using the Vortex SDK.
"""

import sys
import os

# Add parent directory to path to import vortex_sdk
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from vortex_sdk import VortexSdk, RampDirection, FiatToken, EvmToken, Networks


def run_brl_offramp_example():
    """Run the BRL offramp example."""
    try:
        print("Starting BRL Offramp Example...\n")
        
        # Step 1: Initialize VortexSdk
        print("📝 Step 1: Initializing VortexSdk...")
        sdk = VortexSdk(api_base_url="http://localhost:3000")
        print("✅ VortexSdk initialized successfully\n")
        
        # Step 2: Create quote for BRL offramp
        print("📝 Step 2: Creating quote for BRL offramp...")
        quote_request = {
            "from": Networks.POLYGON,
            "inputAmount": "100",
            "inputCurrency": EvmToken.USDC,
            "network": Networks.POLYGON,
            "outputCurrency": FiatToken.BRL,
            "rampType": RampDirection.SELL,
            "to": "pix"
        }
        
        quote = sdk.create_quote(quote_request)
        print("✅ Quote created successfully:")
        print(f"   Quote ID: {quote['id']}")
        print(f"   Input: {quote['inputAmount']} {quote['inputCurrency']}")
        print(f"   Output: {quote['outputAmount']} {quote['outputCurrency']}")
        print(f"   Total Fee: {quote['totalFeeFiat']} {quote['feeCurrency']}")
        print(f"   Expires at: {quote['expiresAt']}\n")
        
        # Step 3: Register the BRL offramp
        print("📝 Step 3: Registering BRL offramp...")
        brl_offramp_data = {
            "pixDestination": "157.492.981-08",
            "receiverTaxId": "157.492.981-08",
            "taxId": "157.492.981-08",
            "walletAddress": "0x1234567890123456789012345678901234567890"
        }
        
        result = sdk.register_ramp(quote, brl_offramp_data)
        ramp_process = result["rampProcess"]
        unsigned_transactions = result["unsignedTransactions"]
        
        print("✅ BRL Offramp registered successfully:")
        print(f"   Ramp ID: {ramp_process['id']}")
        
        # Display unsigned transactions the user must sign
        if unsigned_transactions:
            print("   Unsigned transactions:")
            for tx in unsigned_transactions:
                tx_data = tx.get("txData", {})
                print(f"     - {tx.get('phase')}: Send to {tx_data.get('to')} "
                      f"data {tx_data.get('data')} with value {tx_data.get('value')}")
        print()
        
        # Step 4: User must complete token payment on-chain
        print("🛑 IMPORTANT: Complete the token payment on-chain NOW.")
        print("   Execute the transactions shown above (squidRouterApprove and squidRouterSwap),")
        print("   and save the corresponding transaction hashes.\n")
        
        squid_router_approve_hash = input("➡️  Enter the Squid Router Approve Hash: ").strip()
        squid_router_swap_hash = input("➡️  Enter the Squid Router Swap Hash: ").strip()
        
        # Step 5: Update BRL offramp with transaction hashes
        print("\n📝 Step 4: Updating BRL offramp...")
        transaction_hashes = {
            "squidRouterApproveHash": squid_router_approve_hash,
            "squidRouterSwapHash": squid_router_swap_hash
        }
        
        updated_ramp = sdk.update_ramp(quote, ramp_process["id"], transaction_hashes)
        print("✅ BRL Offramp updated successfully.")
        
        # Step 6: Start the offramp
        print("\n📝 Step 5: Starting BRL offramp...")
        started_ramp = sdk.start_ramp(ramp_process["id"])
        print("✅ Offramp started successfully!")
        print(f"   Current Phase: {started_ramp.get('currentPhase')}")
        
    except Exception as error:
        print(f"❌ Error in BRL Offramp Example: {error}")
        if hasattr(error, '__dict__'):
            print(f"Error details: {error.__dict__}")
        sys.exit(1)


if __name__ == "__main__":
    run_brl_offramp_example()
    print("\n✨ Example execution completed")
    sys.exit(0)
